labeled.go

```diff
@@ -0,0 +1,167 @@
+package ngroklistener
+
+import (
+	"fmt"
+
+	"github.com/caddyserver/caddy/v2"
+	"github.com/caddyserver/caddy/v2/caddyconfig/caddyfile"
+	"go.uber.org/zap"
+	"golang.ngrok.com/ngrok/config"
+)
+
+func init() {
+	caddy.RegisterModule(new(Labeled))
+}
+
+// ngrok Labeled Tunnel
+type Labeled struct {
+	opts []config.LabeledTunnelOption
+
+	// A map of label, value pairs for this tunnel.
+	Labels map[string]string `json:"labels,omitempty"`
+
+	// opaque metadata string for this tunnel.
+	Metadata string `json:"metadata,omitempty"`
+
+	l *zap.Logger
+}
+
+// CaddyModule implements caddy.Module
+func (*Labeled) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{
+		ID: "caddy.listeners.ngrok.tunnels.labeled",
+		New: func() caddy.Module {
+			return new(Labeled)
+		},
+	}
+}
+
+// Provision implements caddy.Provisioner
+func (t *Labeled) Provision(ctx caddy.Context) error {
+	t.l = ctx.Logger()
+
+	if err := t.doReplace(); err != nil {
+		return fmt.Errorf("loading doing replacements: %v", err)
+	}
+
+	if err := t.provisionOpts(); err != nil {
+		return fmt.Errorf("provisioning labeled tunnel opts: %v", err)
+	}
+
+	return nil
+}
+
+func (t *Labeled) provisionOpts() error {
+	for label, value := range t.Labels {
+		t.opts = append(t.opts, config.WithLabel(label, value))
+		t.l.Info("applying label", zap.String("label", label), zap.String("value", value))
+	}
+
+	if t.Metadata != "" {
+		t.opts = append(t.opts, config.WithMetadata(t.Metadata))
+	}
+
+	return nil
+}
+
+func (t *Labeled) doReplace() error {
+	repl := caddy.NewReplacer()
+	replaceableFields := []*string{
+		&t.Metadata,
+	}
+
+	for _, field := range replaceableFields {
+		actual, err := repl.ReplaceOrErr(*field, false, true)
+		if err != nil {
+			return fmt.Errorf("error replacing fields: %v", err)
+		}
+
+		*field = actual
+	}
+
+	// TODO: Labels need to be replaceable
```

Comment:
Yes! You can replace the key and value of the labels. It'll allow for labels such as:
```
label hostname {system.hostname}
```

Reply:
Thanks for pointing out the todo, I meant to come back and implement it. :)